### PR TITLE
DDS-2738, add https pinning, cover with tests

### PR DESCRIPTION
**Pull request by andrevka**

### CODE REVIEW

File: src/Sk/SmartId/Util/Curl.php [hunk 1]
```diff
@@ -325,7 +333,29 @@ public function getError()
...
+  }
+
+  public static function useOnlyLivePublicKey()
+  {
+      self::$server_ssl_public_keys =
+              self::RP_API_PUBLIC_KEY_VALID_FROM_2016_12_20_TO_2020_01_19 . ";"
+              . self::RP_API_PUBLIC_KEY_VALID_FROM_2019_11_01_TO_2021_11_05;
+  }
+
+    public static function setPublicKeysFromArray(array $public_keys)
```
Comment (raigu) [inline src/Sk/SmartId/Util/Curl.php]: Good to see someone is solving SSL pinning issue. Doesn't this approach violate encapsulation? Wouldn't it be more wise to add extra layer so the pinning can be configured through _AuthenticationRequestBuilder_?

Comment (raigu) [reply src/Sk/SmartId/Util/Curl.php]: Hmm, shouldn't the pinning be responsibility of _AuthenticationResponseValidator_?  The downside is that in this way it is not possible to use pinning functionality provided by curl.

Comment (raigu) [reply src/Sk/SmartId/Util/Curl.php]: > Hmm, shouldn't the pinning be responsibility of _AuthenticationResponseValidator_?

Probably not. Digging in code reveals that this is meant to validate data about authenticated user. After the successful authentication server returns user data, and the certificate in it is actually user certificate. So _AuthenticationResponseValidator_ does not have no  idea about server certificate therefore can not validate

**Comment by aasaru** [comment]
@andrevka could you change Curl::useOnlyDemoPublicKey() and others to be non-static?

File: src/Sk/SmartId/Util/Curl.php (hunk 1 above)
Comment (andrevka) [reply src/Sk/SmartId/Util/Curl.php]: > @andrevka could you change Curl::useOnlyDemoPublicKey() and others to be non-static?
Did it

Comment (andrevka) [reply src/Sk/SmartId/Util/Curl.php]: > > Hmm, shouldn't the pinning be responsibility of _AuthenticationResponseValidator_?
> 
> Probably not. Digging in code reveals that this is meant to validate data about authenticated user. After the successful authentication server returns user data which is the input of _\Sk\SmartId\Api\AuthenticationResponseValidator::validate_. The certificate in this input parameter is actually user certificate. Moreover, _AuthenticationResponseValidator_ does not have information about server certificate/public key therefore can not carry out this responsibility anyway.

I think i have resolved the issue, you can check it out.

**Comment by raigu** [comment]
Is it possible to merge methods `useOnlyDemoPublicKey` and `useOnlyLivePublicKey`?

The motivation is to keep test and production code as identical as possible in end application. Then you have the ability to ship a code that is actually tested 100%. Also, applications tend to have a lot of configuration parameters and the less we have them the more productive operations and testing people are. 

I would propose to separate public keys into separate repository (or [pinset](https://github.com/OWASP/CheatSheetSeries/blob/master/cheatsheets/Pinning_Cheat_Sheet.md) in current context) and have the end developer have the possibility to choose how he or she wants to manage pinning.

Example:

```php
# Package will detect suitable keys
$client->usePinning(PublicKeyPinset::detectByUrl($ur));

# Developer can force to use only demo site keys
$client->usePinning(PublicKeyPinset::demoSite());

# Developer can force to use only live site keys
$client->usePinning(PublicKeyPinset::liveSite());

# Developer is free to choose whatever key he or she wants to use.
$keys = array(
    'sha256//QLZIaH7Qx9Rjq3gyznQuNsvwMQb7maC5L4SLu/z5qNU=',
    'whatever format curl\'s CURLOPT_PINNEDPUBLICKEY option accepts'
);

$client->usePinning($keys);
```

In my opinion this makes pinning more flexible and also makes Client interface more lean.

**Comment by jalukse** [comment]
Merged to version 1.5

**Pull request closed without merging**